### ISSUE TITLE
Unable to parse generic types

### ISSUE THREAD

**Issue by marcoabreu** [beanshell/beanshell#22]
I'm using beanshell 2.0b6 to execute regular java-code during the runtime. Unfortunately is the parser unable to parse generic types.

```
for(Map.Entry<String, String> entry : runtime.getParameters().entrySet()) {
}
```

>    Caused by: In file: inline evaluation of: ``package com.marcoabreu.att.aaascripts;    import com.marcoabreu.att.profile.Scri . . . '' Encountered "<" at line 13, column 18.
> 
>   at bsh.Parser.generateParseException(Parser.java:6106)
>   at bsh.Parser.jj_consume_token(Parser.java:5977)
>   at bsh.Parser.EnhancedForStatement(Parser.java:3269)

The same happens if you try to define it as variable:

```
Map.Entry<String, String> entry;
for(entry : runtime.getParameters().entrySet()) {
}
```

> Caused by: In file: inline evaluation of: ``package com.marcoabreu.att.aaascripts;    import com.marcoabreu.att.profile.Scri . . . '' Encountered "," at line 13, column 21.
> 
>   at bsh.Parser.generateParseException(Parser.java:6106)
>   at bsh.Parser.jj_consume_token(Parser.java:5977)
>   at bsh.Parser.Statement(Parser.java:2699)
>   at bsh.Parser.BlockStatement(Parser.java:2819)
>   at bsh.Parser.Block(Parser.java:2779)
>   at bsh.Parser.MethodDeclaration(Parser.java:389)

**Comment by marcoabreu** [comment]
Lambdas dont seem to be supported either.

**Comment by darksidepuffin** [comment]
I've been looking at this on and off. Generics would be useful for my use case.

I'll report back if I manage to sort it out.

**Comment by thirion** [comment]
I found this fork of BeanShell:  https://github.com/pejobo/beanshell2

The following works in that version(2.1.8) but not 2.0b6:
ArrayList<String> entries = new ArrayList(Arrays.asList(new String[]{"hey", "you"}));
for(String entry : entries) {
    print(entry);
}

Any chance of incorporating the functionality from there?

Unfortunately this still doesn't work:
Map.Entry<String, String> entries = runtime.getParameters().entrySet();
for(String entry : entries) {
    print(entry);
}

**Comment by darksidepuffin** [comment]
Never circled back to this - I'm no longer using this in my project.

That said, I'm presuming this isn't supported because BeanShell is apparently based on Java 1.1 grammar, while 1.5/J2SE 5.0 is when Generics were added, and Collections was added in 1.2. 

I never got around to figuring out the grammar parsing/code generation to try to fix this. Someone with more of a vested interest should probably do so.

**Comment by nickl-** [comment]
> Never circled back to this - I'm no longer using this in my project.

No worries, thanx for stopping by much appreciated. I've only recently joined the project and still overwhelmed by the amount of of backlog there is to catch up on. Still I see it as a positive move towards progress and hopefully we can get the community excited about beanshell again. There is a lot of work to be done and we need your help! 

Busy porting/merging with the beanshell2 fork mentioned by @thirion,  incorporating all the fixes/improvements that were made there. I will collect a detailed change log for your reference when I am done.

> That said, I'm presuming this isn't supported because BeanShell is apparently based on Java 1.1 grammar, while 1.5/J2SE 5.0 is when Generics were added, and Collections was added in 1.2.

Already the crew that ran with the bsh2 fork decided on JDK1.5 base and proceeded with ripping out all the archaic hacks, workarounds, junk and replacing it with proper java goodness. I supported that notion and went ahead bumping the source to 1.8. We'll see how that goes down but I have a good feeling though.. things have changed Oracle is driving the bus now. 

In the past there wasn't much reason/interest in upgrading so no one made the effort. The way things are progressing today you want to and will very likely have to upgrade somewhere along the way. Java 7 is end of line, at least you can still build some things with it but my 1.6 is like a white elephant.

Anyway we have to get some excitement going again, even if it is complaining about upgrading JVM =) There is a shell in Java 9 and other projects have reached maturity while bean shell has dragged behind. If we are to still find a place among the giants I don't believe it is going to be an ability to run Noah's VM.

We want generics, right?... and lamda/functional support!!!.. perhaps throw in some EL for syntacticle candy. We should be more JEE aware and what I really want to see is us evolving into proper citizens on the command level. Playing nice with others and utilising the devices that are available in simple, easy and natural ways, as it is supposed to be. Lets see... 

> I never got around to figuring out the grammar parsing/code generation to try to fix this. Someone with more of a vested interest should probably do so.

It seems confusing at first because there's so much backward compatibility cruft in the way which we can hopefully clean up but basically you need your reflection foo on. You could start by writing the test scripts that fail and then go hunting for why they do. Really all these snippets should be in test scripts, how else will we see that it works?

But that's not all there is a LOT of work here, documentation hasn't been updated for years the previous solution is not even available in this fork. There's years of backlog to catch up on, the source forge project needs to be migrated over here, we need a new website, more people to be active in discussions and testing pull requests and and and... if you want to help just go ahead and jump in.

Look forward to seeing more of you...

**Comment by Guardiola31337** [comment]
@nickl- 
> I supported that notion and went ahead bumping the source to 1.8. We'll see how that goes down but I have a good feeling though..

> We want generics, right?

Was generics support enabled on the BeanShell2 fork - [merge-fork-beanshell2 branch](https://github.com/beanshell/beanshell/tree/merge-fork-beanshell2)?

Can't wait to try it out!

Thanks for leading with the `beanshell`'s revival 🙏

**Comment by nickl-** [comment]
Generics now implemented as per #66, parsed and skipped. Raw types are implemented as `java.lang.Object` type classes if name is single character and class not found.
